File: src/io/xol/chunkstories/content/sandbox/GameLogicThread.java
package io.xol.chunkstories.content.sandbox;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import io.xol.chunkstories.Constants;
import io.xol.chunkstories.api.GameContext;
import io.xol.chunkstories.api.GameLogic;
import io.xol.chunkstories.api.plugin.ChunkStoriesPlugin;
import io.xol.chunkstories.api.plugin.PluginManager;
import io.xol.chunkstories.api.plugin.Scheduler;
import io.xol.chunkstories.api.server.Player;
import io.xol.chunkstories.api.world.WorldClient;
import io.xol.chunkstories.client.Client;
import io.xol.chunkstories.content.DefaultPluginManager;
import io.xol.chunkstories.core.events.WorldTickEvent;
import io.xol.chunkstories.server.Server;
import io.xol.chunkstories.tools.ChunkStoriesLogger;
import io.xol.chunkstories.world.WorldClientRemote;
import io.xol.chunkstories.world.WorldImplementation;
import io.xol.chunkstories.world.WorldNetworked;
import io.xol.chunkstories.world.WorldServer;
import io.xol.chunkstories.world.region.RegionImplementation;

//(c) 2015-2016 XolioWare Interactive
//http://chunkstories.xyz
//http://xol.io

/**
 * Sandboxed thread that runs all the game logic for one world, thus including foreign code
 */
public class GameLogicThread extends Thread implements GameLogic
{
	private final GameContext context;
	
	private final WorldImplementation world;
	
	private GameLogicScheduler gameLogicScheduler;
	private boolean die = false;

	public GameLogicThread(WorldImplementation world, SecurityManager securityManager)
	{
		this.world = world;
		this.context = world.getGameContext();
		
		this.setName("World " + world.getWorldInfo().getInternalName()+" logic thread");
		this.setPriority(Constants.MAIN_SINGLEPLAYER_LOGIC_THREAD_PRIORITY);
		
		gameLogicScheduler = new GameLogicScheduler();
		
		//this.start();
	}
	
	public GameContext getGameContext()
	{
		return context;
	}

	long lastNano;
	
	private void nanoCheckStep(int maxNs, String warn)
	{
		long took = System.nanoTime() - lastNano;
		//Took more than n ms ?
		if(took > maxNs * 1000000)
			System.out.println(warn + " " + (double)(took / 1000L) / 1000 + "ms");
		
		lastNano = System.nanoTime();
	}
	
	public void run()
	{
		//Installs a custom SecurityManager
		System.out.println("Security manager: "+System.getSecurityManager());

		while (!die)
		{
			//Dirty performance metric :]
			//perfMetric();
			//nanoCheckStep(20, "Loop was more than 20ms");
			
			//Timings
			fps = 1f / ((System.nanoTime() - lastTimeNs) / 1000f / 1000f / 1000f);
			lastTimeNs = System.nanoTime();
			
			//Updates controller/s views
			if(world instanceof WorldClient)
				Client.getInstance().getClientSideController().updateUsedWorldBits();
			if(world instanceof WorldServer)
			{
				Iterator<Player> i = ((WorldServer)world).getPlayers();
				while(i.hasNext())
				{
					Player p = i.next();
					p.updateUsedWorldBits();
				}
			}
			//nanoCheckStep(1, "World bits");
			
			//Processes incomming pending packets in synch with game logic and flush outgoing ones
			if(world instanceof WorldNetworked)
			{
				((WorldNetworked) world).processIncommingPackets();
				//TODO clean
				if(world instanceof WorldClientRemote)
					((WorldClientRemote) world).getConnection().flush();
				if(world instanceof WorldServer)
					((WorldServer) world).getGameContext().getHandler().flushAll();
			}
			
			//nanoCheckStep(2, "Incomming packets");
			
			this.getPluginsManager().fireEvent(new WorldTickEvent(world));
			
			//Tick the world ( mostly entities )
			world.tick();
			
			//nanoCheckStep(5, "Tick");
			
			//Every second, unloads unused stuff
			if(world.getTicksElapsed() % 60 == 0)
			{
				System.gc();
				
				//Compresses pending chunk summaries
				Iterator<RegionImplementation> loadedChunksHolders = world.getRegionsHolder().getLoadedRegions();
				while(loadedChunksHolders.hasNext())
				{
					RegionImplementation region = loadedChunksHolders.next();
					region.compressChangedChunks();
				}
				
				//Delete unused world data
				world.unloadUselessData();
			}
			
			//nanoCheckStep(1, "unload");
			
			gameLogicScheduler.runScheduledTasks();
			
			//nanoCheckStep(1, "schedule");
			
			//Game logic is 60 ticks/s
			sync(getTargetFps());
		}
	}

	float fps = 0.0f;
	
	long lastTime = 0;
	long lastTimeNs = 0;

	@Override
	public int getTargetFps()
	{
		return 60;
	}

	public double getSimulationFps()
	{
		return (double) (Math.floor(fps * 100f) / 100f);
	}
	
	public double getSimulationSpeed()
	{
		return 1.0;
	}
	
	public void perfMetric()
	{
		double ms = Math.floor((System.nanoTime() - lastTimeNs) / 10000.0) / 100.0;
		String kek = "";
		double d = 0.02;
		for(double i = 0; i < 1.0; i+= d)
			kek += Math.abs(ms - 16 - i) > d ? " " : "|";
		System.out.println(kek + ms + "ms");
	}
	
	public void sync(int fps)
	{
		if (fps <= 0)
			return;

		long errorMargin = 1000 * 1000; // 1 millisecond error margin for
										// Thread.sleep()
		long sleepTime = 1000000000 / fps; // nanoseconds to sleep this frame

		// if smaller than sleepTime burn for errorMargin + remainder micro &
		// nano seconds
		long burnTime = Math.min(sleepTime, errorMargin + sleepTime % (1000 * 1000));

		long overSleep = 0; // time the sleep or burn goes over by

		try
		{
			while (true)
			{
				long t = System.nanoTime() - lastTime;

				if (t < sleepTime - burnTime)
				{
					Thread.sleep(1);
				}
				else if (t < sleepTime)
				{
					// burn the last few CPU cycles to ensure accuracy
					Thread.yield();
				}
				else
				{
					overSleep = Math.min(t - sleepTime, errorMargin);
					break; // exit while loop
				}
			}
		}
		catch (InterruptedException e)
		{
		}

		lastTime = System.nanoTime() - overSleep;
	}

	/* (non-Javadoc)
	 * @see io.xol.chunkstories.content.sandbox.GameLogic#getPluginsManager()
	 */
	@Override
	public PluginManager getPluginsManager()
	{
		return context.getPluginManager();
		/*if(world instanceof WorldClient)
			return Client.getInstance().getPluginManager();
		else if(world instanceof WorldServer)
			return Server.getInstance().getPluginManager();
		return null;*/
	}
	
	public void stopLogicThread()
	{
		die = true;
	}

	@Override
	public Scheduler getScheduler()
	{
		return gameLogicScheduler;
	}
	
	class GameLogicScheduler implements Scheduler {

		List<ScheduledTask> scheduledTasks = new ArrayList<ScheduledTask>();
		
		public void runScheduledTasks()
		{
			try{
			Iterator<ScheduledTask> i = scheduledTasks.iterator();
			while(i.hasNext())
			{
				if(i.next().etc())
					i.remove();
			}
			}
			catch(Throwable t)
			{
				ChunkStoriesLogger.getInstance().error(t.getMessage());
				t.printStackTrace();
			}
		}
		
		@Override
		public void scheduleSyncRepeatingTask(ChunkStoriesPlugin p, Runnable runnable, long delay, long period)
		{
			scheduledTasks.add(new ScheduledTask(runnable, delay, period));
		}
		
		class ScheduledTask {
			
			public ScheduledTask(Runnable runnable, long delay, long period)
			{
				this.runnable = runnable;
				this.delay = delay;
				this.period = period;
			}

			Runnable runnable;
			long delay;
			long period;
			
			//Returns true when it's no longer going to run
			boolean etc()
			{
				if(--delay > 0)
					return false;
				
				runnable.run();
				
				if(period > 0)
					delay = period;
				else
					return true;
				
				return false;
			}
		}
		
	}
}

File: common/src/main/java/xyz/chunkstories/world/logic/GameLogicThread.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.world.logic

import java.util.concurrent.atomic.AtomicBoolean

import org.slf4j.LoggerFactory

import xyz.chunkstories.api.Location
import xyz.chunkstories.api.events.world.WorldTickEvent
import xyz.chunkstories.api.player.entityIfIngame
import xyz.chunkstories.api.util.concurrency.Fence
import xyz.chunkstories.api.world.WorldMaster
import xyz.chunkstories.util.concurrency.SimpleFence
import xyz.chunkstories.util.concurrency.TrivialFence
import xyz.chunkstories.world.WorldImplementation
import java.util.concurrent.ConcurrentLinkedDeque
import java.util.concurrent.Semaphore

abstract class GameLogicThread(private val world: WorldImplementation) : Thread() {
    private val gameLogicScheduler: GameLogicScheduler

    private val die = AtomicBoolean(false)

    private val waitForLogicFinish = SimpleFence()

    internal var lastNano: Long = 0

    internal var fps = 0.0f

    internal var lastTime: Long = 0
    internal var lastTimeNs: Long = 0

    val targetFps: Int
        get() = 60

    val simulationFps: Double
        get() = Math.floor((fps * 100f).toDouble()) / 100f

    val simulationSpeed: Double
        get() = 1.0

    init {
        gameLogicScheduler = GameLogicScheduler()
    }

    private fun nanoCheckStep(maxNs: Int, warn: String) {
        val took = System.nanoTime() - lastNano
        // Took more than n ms ?
        if (took > maxNs * 1000000)
            logger.warn(warn + " " + (took / 1000L).toDouble() / 1000 + "ms")

        lastNano = System.nanoTime()
    }

    override fun run() {
        while (!die.get()) {
            // Dirty performance metric :]
            // perfMetric();
            // nanoCheckStep(20, "Loop was more than 20ms");

            // Timings
            fps = 1f / ((System.nanoTime() - lastTimeNs).toFloat() / 1000f / 1000f / 1000f)
            lastTimeNs = System.nanoTime()

            tick()

            gameLogicScheduler.runScheduledTasks()

            // Game logic is 60 ticks/s
            sync(targetFps)
        }

        waitForLogicFinish.signal()
    }

    fun tickWorld() {
        try {
            world.tick()
        } catch (e: Exception) {
            world.logger.error("Exception occurred while ticking the world : ", e)
        }
    }

    abstract fun tick()

    val lastTimeInMs: Float
        get() = (((System.nanoTime() - lastTimeNs) / 1000.0) / 1000.0).toFloat()

    /** fancy sync method from SO iirc  */
    fun sync(fps: Int) {
        if (fps <= 0)
            return

        val errorMargin = (1000 * 1000).toLong() // 1 millisecond error margin for  Thread.sleep()
        val sleepTime = (1000000000 / fps).toLong() // nanoseconds to sleep this frame

        // if smaller than sleepTime burn for errorMargin + remainder micro &
        // nano seconds
        val burnTime = Math.min(sleepTime, errorMargin + sleepTime % (1000 * 1000))

        var overSleep: Long = 0 // time the sleep or burn goes over by

        try {
            while (true) {
                val t = System.nanoTime() - lastTime

                if (t < sleepTime - burnTime) {
                    Thread.sleep(1)
                } else if (t < sleepTime) {
                    // burn the last few CPU cycles to ensure accuracy
                    Thread.yield()
                } else {
                    overSleep = Math.min(t - sleepTime, errorMargin)
                    break // exit while loop
                }
            }
        } catch (e: InterruptedException) {
        }

        lastTime = System.nanoTime() - overSleep
    }

    fun terminate(): Fence {
        if (!this.isAlive)
            return TrivialFence()

        die.set(true)

        return waitForLogicFinish
    }

    val logicThread: Thread = Thread.currentThread()

    /** Some actions can only execute on the main thread */
    private val actionsQueue = ConcurrentLinkedDeque<ScheduledAction>()

    data class ScheduledAction(val action: () -> Unit) {
        val semaphore = Semaphore(0)
    }

    /** Schedules some work to be executed on the main thread */
    fun logicThread(function: () -> Unit) {
        if (Thread.currentThread() == logicThread) {
            function()
        } else {
            actionsQueue.addLast(ScheduledAction(function))
        }
    }

    fun logicThreadBlocking(function: () -> Unit) {
        if (Thread.currentThread() == logicThread) {
            function()
        } else {
            val scheduled = ScheduledAction(function)
            actionsQueue.addLast(scheduled)
            scheduled.semaphore.acquireUninterruptibly()
        }
    }
    companion object {
        private val logger = LoggerFactory.getLogger("gameLogic")
    }
}
